File: ModelIO_Bindings/ModelIO_Bindings/ModelIOAllocators.swift
//
//  ModelIOAllocators.swift
//  ModelIO_Bindings
//
//  Created by Taanish Reja on 1/8/25.
//

import Metal
import MetalKit
import ModelIO

public func createMTKMeshBufferAllocator(devicePtr: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer
{
    let device = Unmanaged<MTLDevice>.fromOpaque(devicePtr).takeUnretainedValue();
    
    let mtkMeshBufferAllocator = MTKMeshBufferAllocator(device: device);
    
    let mtkMeshBufferAllocatorPtr = UnsafeMutableRawPointer(Unmanaged.passRetained(mtkMeshBufferAllocator).toOpaque());
    
    return mtkMeshBufferAllocatorPtr;
}

public func releaseMTKMeshBufferAllocator(mtkMeshBufferAllocatorPtr: UnsafeMutableRawPointer)
{
    let mtkMeshBufferAllocator = Unmanaged<MTKMeshBufferAllocator>.fromOpaque(mtkMeshBufferAllocatorPtr);
    mtkMeshBufferAllocator.release();
}

File: ModelIO_Bindings/ModelIO_Bindings/ModelIOAllocators.cpp
//
//  ModelIOAllocators.cpp
//  ModelIO_Bindings
//
//  Created by Taanish Reja on 1/9/25.
//

#include "ModelIOAllocators.h"

namespace ModelIO {
    MTKMeshBufferAllocator::MTKMeshBufferAllocator(MTL::Device* device):
        m_allocator{std::make_shared<SwiftObject>(
            ModelIO_Bindings::createMTKMeshBufferAllocator(static_cast<void*>(device)),
            ModelIO_Bindings::releaseMTKMeshBufferAllocator
        )}
    {}

    MTKMeshBufferAllocator::~MTKMeshBufferAllocator() = default;

    SwiftObject* MTKMeshBufferAllocator::getAllocator() {
        return m_allocator.get();
    }
}

